src/context_free.ml

```diff
@@ -518,18 +518,30 @@ class map_top_down ?(expect_mismatch_handler = Expect_mismatch_handler.nop)
           | None ->
               self#pexp_apply_without_traversing_function base_ctxt e func args
           | Some pattern -> (
-              match pattern e with
+              let pat_expr =
```

Comment:
`pattern` was already a strange name for a function transforming an expression. Let's not use `pat_expr` as name for the generated code, but something more explicit such as `generated_code`.